File: examples/plugins/react-native-idfa-plugin/ios/IdfaPlugin.m
#import <Foundation/Foundation.h>
#import "React/RCTBridgeModule.h"

@interface RCT_EXTERN_MODULE(IdfaPlugin, NSObject)

RCT_EXTERN_METHOD(requestTrackingAuthorization: (RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getIdfa: (RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

@end

File: examples/plugins/react-native-idfa-plugin/ios/IdfaPlugin.swift
import AdSupport
import AppTrackingTransparency
import Foundation

@objc(IdfaPlugin)
class IdfaPlugin: NSObject {
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc
  func requestTrackingAuthorization(
      _ resolve: @escaping RCTPromiseResolveBlock,
      rejecter reject: RCTPromiseRejectBlock
  ) -> Void {
    if #available(iOS 14, *) {
      ATTrackingManager.requestTrackingAuthorization { status in
        resolve(status.rawValue)
      }
    } else {
      resolve(0)
    }
  }

  @objc
  func getIdfa(
      _ resolve: @escaping RCTPromiseResolveBlock,
      rejecter reject: RCTPromiseRejectBlock
  ) -> Void {
    if #available(iOS 14, *) {
      let status = ATTrackingManager.trackingAuthorizationStatus
      if (status == ATTrackingManager.AuthorizationStatus.authorized) {
        resolve(ASIdentifierManager.shared().advertisingIdentifier.uuidString)
      } else {
        resolve(nil)
      }
    } else {
      resolve(ASIdentifierManager.shared().advertisingIdentifier.uuidString)
    }
  }
}
